### PR TITLE
Sticky Blocks

### PR DESCRIPTION
**Pull request by HardeepAsrani**

### CODE REVIEW

**Comment by Soare-Robert-Daniel** [comment]
## Testing

To make a block sticky, open the block menu and select the option. Same step for removing.

image


## Options

#### Stick To
This select the container in which the element will be sticky

:warning: On the frontend, you can write in the console `debugSticky = true` to see the boundaries.

| Value | Action |
| -- | -- |
| Default | The container will be the main Section/Group/Columns in the List View hierarchy |
| Parent Column/Group | The container will be the direct parent, which can be a Column/Group child block |
| Section | The container will be Section/Group/Columns, that directly contains the sticky block |
| Screen | The container will the screen/window |

stick-heir


#### Position
| Value | Action |
| -- | -- |
| Top | Stick to the top of the screen |
| Bottom | Stick to the bottom of the screen |

#### Offset
This is the distance from the screen to the sticky element.

#### Behaviour
Select what happens when sticky elements collide with one another.

:warning: This is only for sticky elements with the same container. Work best with `Default` and `Screen`. For granular selections like `Parent Column/Group` and `Section`, extra attention is needed.

| Value | Action |
| -- | -- |
| Default | The sticky element will go behind the next sticky element |
| Hide | The sticky element will fade out when approaching the next sticky element |
| Stack | The sticky element will position itself after the previous one |

:warning: Stack is working only with sticky elements that have the same option and the same container. Work best with `Default` and `Screen`. 

## Testing Templates
[template6-multiple-section.txt](https://github.com/Codeinwp/otter-blocks/files/7763309/template6-multiple-section.txt)
[template5-subscribe.txt](https://github.com/Codeinwp/otter-blocks/files/7763310/template5-subscribe.txt)
[template4-sidebar.txt](https://github.com/Codeinwp/otter-blocks/files/7763311/template4-sidebar.txt)
[template3-ads.txt](https://github.com/Codeinwp/otter-blocks/files/7763312/template3-ads.txt)
[template2-stack-option.txt](https://github.com/Codeinwp/otter-blocks/files/7763313/template2-stack-option.txt)

-----
@irinelenache, @HardeepAsrani, when testing, please post the video with the problem, the screenshots do not help too much. Also activated the debug mode with `debugSticky = true`.

⚠️ The plugin is sensible to screen resize, if you change the width of the windows, please refresh the page.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested the feature and found some things:

- [ ] With [these settings](https://vertis.d.pr/sExEp4) for a **section**, the sticky block goes over the header https://vertis.d.pr/0hoqv0 . I think that when i scroll back to the top, the block should stay in his initial position
- [ ] With [these settings](https://vertis.d.pr/EHyZ6k) for a **column**, the block becomes sticky after scrolling over it https://vertis.d.pr/GawWir
- [ ] When a block is sticked to **Parent column** or **Section**, there are 2 scrollbars in the Chrome window https://vertis.d.pr/xU6Oi4
- [ ] With [these settings](https://vertis.d.pr/UkHmRS) for a block that's sticked to **Parent column**, it behaves like this in frontend https://vertis.d.pr/y8bfQF
- [ ] With 3 sticky blocks (a section, a column and a simple block), all have [the same settings](https://vertis.d.pr/N23Tv5), a block starts to fade after it passed the next one, i don't know if this is expected https://vertis.d.pr/disiAb

Please let me know if i didn't exactly understand the feature 🙏

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, you can re-test the PR. 

> * [x]  With [these settings](https://vertis.d.pr/sExEp4) for a **section**, the sticky block goes over the header https://vertis.d.pr/0hoqv0 . I think that when i scroll back to the top, the block should stay in his initial position

I have add a limit for the elements with position `Bottom` that are at the beginning of the screen. You will see an updated information in inspector for Position. Also you will get a warning in the console on frontend when an element does not meet the limit.

> * [x]  With [these settings](https://vertis.d.pr/EHyZ6k) for a **column**, the block becomes sticky after scrolling over it https://vertis.d.pr/GawWir
> * [x]  When a block is sticked to **Parent column** or **Section**, there are 2 scrollbars in the Chrome window https://vertis.d.pr/xU6Oi4
> * [x]  With [these settings](https://vertis.d.pr/UkHmRS) for a block that's sticked to **Parent column**, it behaves like this in frontend https://vertis.d.pr/y8bfQF

This was caused by `body` which had a fixed `height`. I have added an exception for this.

> * [x]  With 3 sticky blocks (a section, a column and a simple block), all have [the same settings](https://vertis.d.pr/N23Tv5), a block starts to fade after it passed the next one, i don't know if this is expected https://vertis.d.pr/disiAb

It is expected in this case. There elements with the behavior set to `Hide` and with the same container (in this case `Screen`)

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, now the sticky elements will change their horizontal position when resizing the window.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel The reported issues are fixed, thanks 🚀 

Found something, let me know if it can be fixed or it is expected:

- [ ] With [these settings](https://vertis.d.pr/uU5MvV) for a Column, there is a strange behaviour when i reach the sticky element https://vertis.d.pr/E81GSz

**Comment by Soare-Robert-Daniel** [comment]
> * [ ]  With [these settings](https://vertis.d.pr/uU5MvV) for a Column, there is a strange behaviour when i reach the sticky element https://vertis.d.pr/E81GSz

It is expected since you have a big `offset` value. Offset also acts as a trigger limit.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel If I choose "Parent" option and the element has no Columns as a parent, what will happen?

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani it will use the screen as a parent. The fallback for all options is the `screen` option. If you use sticky an element that does not have a parent, it will be the entire screen.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel That is confusing. We should not show those options if that is the case. If there is no parent Section, let's just disable those options.

And it should also support Core's Column/Group block.

File: inc/class-registration.php [hunk 1]
```diff
@@ -492,6 +492,18 @@ public function enqueue_dependencies( $post = null ) {
 		if ( ! self::$scripts_loaded['product-image'] && has_block( 'themeisle-blocks/product-image', $post ) ) {
 			wp_enqueue_script( 'wc-single-product' );
 		}
+
+		// DEBUG
+		// TODO: load this only when a block has a sticky block.
+		$asset_file = include OTTER_BLOCKS_PATH . '/build/blocks/sticky.asset.php';
+		wp_enqueue_script(
```
Comment (Soare-Robert-Daniel) [inline inc/class-registration.php:499]: Any idea on how to load this only if sticky is present? Should we analyze the `className` attr?

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Looks good now. Just one thing, can you improve the field description to something better?

Like:

> Set the distance from the block to the screen when sticky mode is active. 

It can just be:

> Distance from the block to the screen.

Or even

> Make the sticky mode active for mobile users. Please check the functionality in Preview mode, some elements might not work properly

Don't think it needs to be there. That toggle is self-descriptive.

And so on, they're long and confusing. We can try to phrase them better.

**Pull request merged**